pkg/compactor/block_upload.go

```diff
@@ -104,6 +104,15 @@ func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.
 	}
 
 	if c.cfgProvider.CompactorBlockUploadValidationEnabled(tenantID) {
+		maxConcurrency := int64(c.compactorCfg.MaxBlockUploadValidationConcurrency)
+		if maxConcurrency > 0 && c.blockUploadValidations.Load() >= maxConcurrency {
```

Comment:
Trusting `c.blockUploadValidations.Load()` allows a race where two can enter at once before either increments. Peeking with a `Load()` first is fine, but the actual check should `Inc()` and check if the return is <= maxConcurrency. If that check fails a decrement must follow though.

Reply:
Here's an example from ingester push method: https://github.com/grafana/mimir/blob/cfd882b23714deddaf71398e326ed9ff3d2611b5/pkg/ingester/ingester.go#L659-L668